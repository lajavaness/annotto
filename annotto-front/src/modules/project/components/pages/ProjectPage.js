import { isEmpty, merge, omit } from 'lodash'
import { message } from 'antd'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import React, { useCallback, useEffect, useMemo } from 'react'

import {
	selectIsFilterModalOpen,
	selectIsGuideModalOpen,
	selectIsReady,
	selectIsSuccessPosted,
	selectProjectAdmins,
	selectProjectDatascientists,
	selectProjectDefaultTags,
	selectProjectEntitiesRelationsGroup,
	selectProjectFilter,
	selectProjectFilterFields,
	selectProjectFilterOperators,
	selectProjectGuideLines,
	selectProjectId,
	selectProjectItemsTags,
	selectProjectTasks,
	selectProjectUsers,
} from 'modules/project/selectors/projectSelectors'
import { selectIsSimilarityAllowed } from 'modules/root/selectors/rootSelectors'

import { closeFilterModal, closeGuideModal } from 'modules/project/actions/projectActions'
import { postProjectFilter } from 'modules/project/actions/filterActions'

import FilterForm from 'modules/project/components/common/FilterForm'
import GuideModal from 'modules/project/components/common/GuideModal'
import Loader from 'shared/components/loader/Loader'
import Page from 'shared/components/page/Page'

import * as Styled from 'modules/project/components/pages/__styles__/ProjectPage.styles'

const ProjectPage = ({ children }) => {
	const isReady = useSelector(selectIsReady)
	const filter = useSelector(selectProjectFilter)
	const projectId = useSelector(selectProjectId)
	const isSuccessPosted = useSelector(selectIsSuccessPosted)
	const filterOperators = useSelector(selectProjectFilterOperators)
	const filterFields = useSelector(selectProjectFilterFields)
	const isFilterModalOpen = useSelector(selectIsFilterModalOpen)
	const isGuideModalOpen = useSelector(selectIsGuideModalOpen)
	const guidelines = useSelector(selectProjectGuideLines)
	const tasks = useSelector(selectProjectTasks)
	const entitiesRelationsGroup = useSelector(selectProjectEntitiesRelationsGroup)
	const users = useSelector(selectProjectUsers)
	const datascientists = useSelector(selectProjectDatascientists)
	const admins = useSelector(selectProjectAdmins)
	const isSimilarityAllowed = useSelector(selectIsSimilarityAllowed)
	const defaultTags = useSelector(selectProjectDefaultTags)
	const itemTags = useSelector(selectProjectItemsTags)

	const dispatch = useDispatch()

	const { t } = useTranslation('project')

	useEffect(() => {
		if (isSuccessPosted) {
			dispatch(closeFilterModal())
			message.success(t('project:filter.successMessage'))
		}
	}, [isSuccessPosted, dispatch, t])

	const tags = merge(defaultTags, itemTags)

	const sortedUsers = useMemo(
		() =>
			[
				...(!isEmpty(users) ? users : []),
				...(!isEmpty(admins) ? admins : []),
				...(!isEmpty(datascientists) ? datascientists : []),
			].sort(),
		[admins, datascientists, users]
	)

	const _onFilterModalCloseClick = useCallback(() => dispatch(closeFilterModal()), [dispatch])

	const _onGuideModalCloseClick = useCallback(() => dispatch(closeGuideModal()), [dispatch])

	const _onSubmitFilterClick = useCallback(
		(value) => dispatch(postProjectFilter(projectId, value)),
		[dispatch, projectId]
	)

	const _onCancelClick = useCallback(() => dispatch(closeFilterModal()), [dispatch])

	return (
		<Page id={'project'}>
			{isReady ? (
				<Styled.Root>
					{children}
					<Styled.Modal
						width={'100%'}
						destroyOnClose={true}
						getContainer={false}
						footer={null}
						visible={isFilterModalOpen}
						onCancel={_onFilterModalCloseClick}
					>
						<FilterForm
							tags={tags}
							formData={filter?.form}
							isSimilarityAllowed={isSimilarityAllowed}
							filterOperators={filterOperators}
							filterFields={isSimilarityAllowed ? filterFields : omit(filterFields, 'uuid')}
							users={sortedUsers}
							tasks={tasks}
							onSubmitFilterClick={_onSubmitFilterClick}
							onCancelClick={_onCancelClick}
						/>
					</Styled.Modal>
					<Styled.Modal
						width={'100%'}
						destroyOnClose={true}
						getContainer={false}
						footer={null}
						visible={isGuideModalOpen}
						onCancel={_onGuideModalCloseClick}
					>
						<GuideModal guidelines={guidelines} tasks={tasks} entitiesRelationsGroup={entitiesRelationsGroup} />
					</Styled.Modal>
				</Styled.Root>
			) : (
				<Loader />
			)}
		</Page>
	)
}

export default ProjectPage
