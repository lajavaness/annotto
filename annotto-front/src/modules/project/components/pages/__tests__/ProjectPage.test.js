import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import ProjectPage from 'modules/project/components/pages/ProjectPage'

const initialState = {}
const getStore = (state) => createStore({ initialState: state })
const getInstance = (state = initialState) => (
	<Provider store={getStore(state)}>
		<ThemeProvider theme={theme}>
			<ProjectPage />
		</ThemeProvider>
	</Provider>
)

describe('ProjectPage', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
