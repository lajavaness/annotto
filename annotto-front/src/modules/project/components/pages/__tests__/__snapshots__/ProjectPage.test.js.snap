// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPage matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-dkaWRx heWsvl"
    data-testid="__project-page__"
  >
    <div
      class="sc-eisiLS JIdvx"
      data-testid="__loader__"
    >
      <div
        class="ant-spin ant-spin-lg ant-spin-spinning"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
